cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/FlowControlMonitorService.java

```diff
@@ -0,0 +1,189 @@
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.services;
+
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.app.program.ProgramDescriptor;
+import io.cdap.cdap.app.runtime.ProgramOptions;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.store.AppMetadataStore;
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Maintain and provides total number of launching and running run-records. This class is used by
+ * flow-control mechanism for launch requests.
+ */
+public class FlowControlMonitorService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(FlowControlMonitorService.class);
+
+  // Program types controlled by flow-control mechanism.
+  private static final Set<ProgramType> CONTROL_FLOW_PROGRAM_TYPES = EnumSet.of(ProgramType.MAPREDUCE,
+      ProgramType.WORKFLOW,
+      ProgramType.SPARK,
+      ProgramType.WORKER);
+
+  private final MetricsCollectionService metricsCollectionService;
+  private final TransactionRunner transactionRunner;
+
+  /**
+   * Monitors the program flow control.
+   *
+   * @param metricsCollectionService collect metrics
+   */
+  @Inject
+  public FlowControlMonitorService(
+      MetricsCollectionService metricsCollectionService,
+      TransactionRunner transactionRunner) {
+    this.metricsCollectionService = metricsCollectionService;
+    this.transactionRunner = transactionRunner;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("FlowControlMonitorService started.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("FlowControlMonitorService successfully shut down.");
+  }
+
+  /**
+   * Add a new in-flight launch request and return total number of launching and running programs.
+   *
+   * @param programRunId run id associated with the launch request
+   * @return total number of launching and running program runs.
+   */
+  public Counter addRequestAndGetCount(ProgramRunId programRunId, ProgramOptions programOptions,
+      ProgramDescriptor programDescriptor) throws Exception {
+    if (RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS) == -1) {
+      throw new Exception("None time-based UUIDs are not supported");
+    }
+
+    int launchingCount = addRequest(programRunId, programOptions, programDescriptor);
+    int runningCount = getFlowControlMetrics(false, true).getRunningCount();
```

Comment:
Are there value in having `getFlowControlMetrics` method? in `addRequestAndGetCount` it still ends up creating 2 separate database transactions. I'd consider moving transaction into addRequestAndGetCount/getCount and removing/splitting getFlowControlMetrics into 2 methods without transaction handling.
As of emitFlowControlMetrics, as far as I can see in all but edge case all metrics are emitted, so I'd just make it emit both metrics universaly.

Reply:
Refactored code and merged methods in one transaction.

* Merged `getFlowControlMetrics()` and `getCount()` -> I feel it is useful to get both the counters or either of them.
* Added store queries in one transaction in `addRequestAndGetCount()` and `addRequest()`.
* emitFlowControlMetrics can continue to emit the metrics based on the parameters.

Reply:
What's the value of `emit based on parameters`? There is only one usage where parameters are not `true, true` and that's an edge case (program running failed). I'd just make it to emit both always. It's less error prone. We already have some bug where we underemitted metrics

Reply:
Made the change for `emitFlowControlMetrics()` and `getCount()`. Removed the parameters.